### PR TITLE
LoadSampleShape that reads Ascii OFF files

### LINKED ISSUE #21553: LoadSampleShape algorithm extended to support OFF Geometry format
Make the LoadShape algorithm in #12828 also support loading the another triangular based mesh geometry format, here the OFF format. https://en.wikipedia.org/wiki/OFF_(file_format)

In this issue make LoadShape support the know wikipedia format. A separate will be created to handle the new proposed nexus OFF format.

### PR DESCRIPTION
**Pull request by KarlPalmen**
**Description of work**
Extend LoadSampleShape to work with ASCII OFF files as described at 
https://en.wikipedia.org/wiki/OFF_(file_format) 

**To test:**
Run MantidPlot to obtain a workspace with an instrument and then run LoadSampleShape on one of the
[off.zip](https://github.com/mantidproject/mantid/files/1974044/off.zip) files attached or a modification of such a file made by you.

I found some more OFF files at
http://people.sc.fsu.edu/~jburkardt/data/off/off.html
Of these `cone.off, icosa.off, mctet.off, mctri.off, octa.off, socket.off & tetra.off` have triangular faces, so could be used for testing.

Fixes #21553 . 

**Release Notes** 
See https://github.com/mantidproject/mantid/pull/22385/commits/0c4a3f6bfa36811a293e7da2ec5f3984cd4cb765

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by NickDraper** [comment]
@OwenArnold I've had a quick once over of this but I'd appreciate you taking a look as well

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 1]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+      }
+      vertices.push_back(vertex);
+    } else {
+      throw std::runtime_error(
+          "Unexpected end of file, while reading OFF vertices");
+    }
+  }
+}
+
+void readOFFTriangles(std::ifstream &file, uint16_t nT,
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: you might as well spell `numTriangles` as opposed to `nT` (minor)

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 2]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+  uint16_t nVertices;
+  uint16_t nTriangles;
+
+  std::string line;
+  // Get number of vetrtices and faces
+  if (getOFFline(file, line)) {
+    std::vector<std::string> tokens;
+    boost::split(tokens, line, boost::is_any_of(" "), boost::token_compress_on);
+    if (tokens.size() == 3) {
+      nVertices = boost::lexical_cast<uint16_t>(tokens[0]);
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: What happens when an instrument geometry has more than (2^16) 65536 vertices? Some ESS instruments will have more than 100K detectors which could potentially be defined as OFF files.

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 3]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+  while (line.empty() || line.substr(0, 1) == "#") {
+    if (!getline(file, line)) {
+      return false;
+    }
+    boost::trim(line);
+  }
+  return true;
+}
+
+void readOFFVertices(std::ifstream &file, uint16_t nV,
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: same as for `nT` but minor.

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 4]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+                     std::vector<V3D> &vertices) {
+  std::string line;
+  V3D vertex;
+  for (uint16_t i = 0; i < nV; i++) {
+    if (getOFFline(file, line)) {
+      std::vector<std::string> tokens;
+      boost::split(tokens, line, boost::is_any_of(" "),
+                   boost::token_compress_on);
+      if (tokens.size() == 3) {
+        vertex.setX(boost::lexical_cast<double>(tokens[0]));
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: could probably just emplace into `vertices` here without using `vertex`. You'll save on all the copies for large numbers of vertices.
```
vertices.emplace_back(boost::lexical_cast<double>(tokens[0]), //x
                      boost::lexical_cast<double>(tokens[1]), //y
                      boost::lexical_cast<double>(tokens[2])) //z
````

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 5]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+  size_t nFaceVertices;
+  for (uint16_t i = 0; i < nT; i++) {
+    if (getOFFline(file, line)) {
+      std::vector<std::string> tokens;
+      boost::split(tokens, line, boost::is_any_of(" "),
+                   boost::token_compress_on);
+      if (tokens.size() >= 4) {
+        nFaceVertices = boost::lexical_cast<size_t>(tokens[0]);
+        if (nFaceVertices == 3) {
+          t1 = boost::lexical_cast<uint16_t>(tokens[1]);
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: again, what happens when vertex indices exceed 2^16?

File: Framework/DataHandling/src/LoadSampleShape.cpp [hunk 6]
```diff
@@ -140,6 +140,123 @@ std::unique_ptr<Geometry::MeshObject> readSTLSolid(std::ifstream &file,
...
+          t2 = boost::lexical_cast<uint16_t>(tokens[2]);
+          t3 = boost::lexical_cast<uint16_t>(tokens[3]);
+        } else {
+          throw std::runtime_error("OFF face is not a triangle.");
+        }
+        triangleIndices.push_back(t1);
```
Comment (LamarMoore) [inline Framework/DataHandling/src/LoadSampleShape.cpp]: emplace?

**Review by LamarMoore — CHANGES_REQUESTED** [review]
Looks good overall. Good testing. Just did a code review, haven't done a functional test yet. A few minor points below.

File: docs/source/release/v3.13.0/framework.rst [hunk 7]
```diff
@@ -22,7 +22,7 @@ New Algorithms
 ##############
 
 - :ref:`LoadSampleShape <algm-LoadSampleShape>` loads a shape into the sample in a workspace from an STL file,
-  which contains a list of triangles.
+  which contains a list of triangles or an OFF file that has a list of vertices and triangles.
```
[the remark this answers is not in the document]
Comment (KarlPalmen) [reply docs/source/release/v3.13.0/framework.rst]: and better still also a link to the STL wikipedia definition.

File: Framework/DataHandling/src/LoadSampleShape.cpp (hunk 2 above)
Comment (LamarMoore) [reply Framework/DataHandling/src/LoadSampleShape.cpp]: The OFF format is currently being added as a way to represent instrument geometry by the Nexus Standard. I am currently working on a fork of Mantid which introduces OFF geometry using `MeshObject` so this change is imminent.

Comment (LamarMoore) [reply Framework/DataHandling/src/LoadSampleShape.cpp]: I am happy for that to be handled in a separate PR though. Doesn't have to be dealt with here and now.

Comment (KarlPalmen) [reply Framework/DataHandling/src/LoadSampleShape.cpp]: If you believe there will be MeshObjects with more than 2^16 vertices, please raise it as an issue or make the necessary changes yourself. If you do make the necessary changes, LoadSampleShape would need to be modified after this issue merged with master.

File: Framework/DataHandling/src/LoadSampleShape.cpp (hunk 5 above)
Comment (KarlPalmen) [reply Framework/DataHandling/src/LoadSampleShape.cpp]: If execution gets this far the vertex index would be out of range.

**Comment by KarlPalmen** [comment]
@LamarMoore I have addressed your review.

**Review by LamarMoore — APPROVED** [review]

**Pull request merged**